### ISSUE TITLE
Generator does not stop in spite of RuntimeException thrown into parent scope

### ISSUE THREAD

**Issue by mpyw** [mpyw/co#11]
```php
<?php
require __DIR__ . '/../../vendor/autoload.php';

use mpyw\Co\Co;
use mpyw\Co\CURLException;

class TerminatedException extends \RuntimeException {}

Co::wait(function () {
    try {
        yield [timer(), main()];
    } catch (TerminatedException $e) {
        var_dump('Terminated.');
    }
});

function curl_init_with($url, array $options = [CURLOPT_RETURNTRANSFER => true])
{
    $ch = curl_init($url);
    curl_setopt_array($ch, $options);
    return $ch;
}

function timer()
{
    $ms = 0;
    while (true) {
        yield CO::DELAY => 0.2;
        $ms += 200;
        echo "[Timer]: $ms miliseconds passed\n";
    }
}

function main()
{
    var_dump(array_map('strlen', yield [
        'Content-Length of github.com' => curl_init_with('https://github.com/mpyw'),
        'Content-Length of twitter.com' => curl_init_with('https://twitter.com/mpyw'),
    ]));
    throw new TerminatedException;
}
```

will output...

```ShellSession
mpyw@localhost:~$ php examples/timer/request-timer.php
[Timer]: 200 miliseconds passed
[Timer]: 400 miliseconds passed
[Timer]: 600 miliseconds passed
[Timer]: 800 miliseconds passed
[Timer]: 1000 miliseconds passed
[Timer]: 1200 miliseconds passed
[Timer]: 1400 miliseconds passed
[Timer]: 1600 miliseconds passed
array(2) {
  ["Content-Length of github.com"]=>
  int(80324)
  ["Content-Length of twitter.com"]=>
  int(279605)
}
string(11) "Terminated."
[Timer]: 1800 miliseconds passed
[Timer]: 2000 miliseconds passed
[Timer]: 2200 miliseconds passed
[Timer]: 2400 miliseconds passed
[Timer]: 2600 miliseconds passed
[Timer]: 2800 miliseconds passed
[Timer]: 3000 miliseconds passed
[Timer]: 3200 miliseconds passed
[Timer]: 3400 miliseconds passed
[Timer]: 3600 miliseconds passed
[Timer]: 3800 miliseconds passed
...
```

**Comment by mpyw** [comment]
This is not a `Co` bug, but a design pattern issue.
ES2015 Promise also performs like this.

```html
<!DOCTYPE html>
<meta charset="UTF-8">
<title>Example</title>
<script async src="https://cdnjs.cloudflare.com/ajax/libs/co/4.1.0/index.min.js"></script>
<script async src="https://code.jquery.com/jquery-2.2.2.min.js"></script>
<script>
    'use strict';

    new Promise(r => addEventListener('DOMContentLoaded', r))
    .then(() => Promise.all(Array.from(document.querySelectorAll('script[async]')).map(
        e => new Promise(r => e.addEventListener('load', r))
    )))
    .then(() => co(main))
    .catch(e => console.error(e));

    function* main() {
        try {
            yield [timer(), request()];
        } catch (e) {
            console.error(e);
        }
    }
    
    function sleep(ms) {
        return new Promise(r => setTimeout(r, ms));
    }
    
    function* timer() {
        let ms = 0;
        while (true) {
            yield sleep(200);
            ms += 200;
            console.log('[Timer]: ' + ms + ' miliseconds passed');
        }
    }
    
    function* request() {
        let q = 'http://news.google.com/news?hl=ja&ned=us&ie=UTF-8&oe=UTF-8&output=atom&topic=h';
        let url = 'https://ajax.googleapis.com/ajax/services/feed/load?v=1.0&num=10&q=' + encodeURIComponent(q);
        let entries = (yield $.ajax(url, {
            dataType: 'jsonp',
            jsonpCallBack: 'callback',
        })).responseData.feed.entries;
        console.log(entries);
        throw new Error('Terminated.')
    }
</script>
```

will output...

![2016-07-20 17 30 57](https://cloud.githubusercontent.com/assets/1351893/16980193/bf81eb0c-4e9f-11e6-96ce-1db12eee23a3.png)

**Comment by mpyw** [comment]
The right way

```php
<?php
require __DIR__ . '/../../vendor/autoload.php';

use mpyw\Co\Co;
use mpyw\Co\CURLException;

class TerminatedException extends \RuntimeException {}

Co::wait(function () {
    try {
        yield [timer($e), main()];
    } catch (TerminatedException $_) {
        $e = $_; // Since PHP Bug #72629, we need to assign caught exception to another variable.
        var_dump('Terminated.');
    }
});

function curl_init_with($url, array $options = [CURLOPT_RETURNTRANSFER => true])
{
    $ch = curl_init($url);
    curl_setopt_array($ch, $options);
    return $ch;
}

function timer(&$e)
{
    $ms = 0;
    while (true) {
        yield CO::DELAY => 0.2;
        if ($e) {
            return;
        }
        $ms += 200;
        echo "[Timer]: $ms miliseconds passed\n";
    }
}

function main()
{
    var_dump(array_map('strlen', yield [
        'Content-Length of github.com' => curl_init_with('https://github.com/mpyw'),
        'Content-Length of twitter.com' => curl_init_with('https://twitter.com/mpyw'),
    ]));
    throw new TerminatedException;
}
```

**Comment by mpyw** [comment]
Reopened this bug because uncaught exception should still stop all jobs.

**Comment by mpyw** [comment]
In fact, we don't need to throw exceptions on the purpose of breaking infinite loops.
Better codes are now available: 7505c5c
